connectors/src/connectors/webcrawler/temporal/activities.ts

```diff
@@ -67,68 +72,106 @@ export async function crawlWebsiteByConnectorId(connectorId: ModelId) {
+                    connectorId: connector.id,
+                    webcrawlerConfigurationId: webCrawlerConfig.id,
+                    url: {
+                      [Op.in]: Array.from(pagesForFolder),
+                    },
+                  },
+                }
+              );
+            }
           }
+          const documentId = stableIdForUrl({
+            url: request.url,
+            ressourceType: "file",
+          });
 
-          const logicalParent =
-            folder === webCrawlerConfig.url ? null : getFolderForUrl(folder);
-          await WebCrawlerFolder.upsert({
-            url: folder,
+          let logicalParent: string | null = null;
+          if (!isTopFolder(request.url)) {
+            const normalizedPageUrl = normalizeFolderUrl(request.url);
+            if (createdFolders.has(normalizedPageUrl)) {
+              logicalParent = normalizedPageUrl;
+            } else {
+              logicalParent = getFolderForUrl(request.url);
+            }
+          }
+          await WebCrawlerPage.upsert({
+            url: request.url,
             parentUrl: logicalParent,
             connectorId: connector.id,
             webcrawlerConfigurationId: webCrawlerConfig.id,
-            internalId: stableIdForUrl({
-              url: folder,
-              ressourceType: "folder",
-            }),
+            documentId: documentId,
+            title: pageTitle,
           });
-          createdFolders.add(folder);
-        }
-        const documentId = stableIdForUrl({
-          url: request.url,
-          ressourceType: "file",
-        });
-        const logicalParent =
-          request.url === webCrawlerConfig.url
-            ? null
-            : getFolderForUrl(request.url);
-        await WebCrawlerPage.upsert({
-          url: request.url,
-          parentUrl: logicalParent,
-          connectorId: connector.id,
-          webcrawlerConfigurationId: webCrawlerConfig.id,
-          documentId: documentId,
-          title: pageTitle,
-        });
 
-        await upsertToDatasource({
-          dataSourceConfig,
-          documentId: documentId,
-          documentContent: {
-            prefix: pageTitle,
-            content: extracted,
-            sections: [],
-          },
-          documentUrl: request.url,
-          timestampMs: new Date().getTime(),
-          tags: [`title:${pageTitle.substring(0, 300)}`],
-          parents: [documentId].concat(
-            folders.map((x) =>
-              stableIdForUrl({ url: x, ressourceType: "folder" })
-            )
-          ),
-          upsertContext: {
-            sync_type: "batch",
-          },
-        });
+          const normalizedPageUrl = normalizeFolderUrl(request.url);
+          if (!page2normalizedFolder.has(normalizedPageUrl)) {
+            page2normalizedFolder.set(normalizedPageUrl, new Set());
+          }
+          page2normalizedFolder.get(normalizedPageUrl)?.add(request.url);
+
+          await upsertToDatasource({
```

Comment:
We likely don't want to serialize calls to upsertToDatasource ?

Should we instead put an advisory lock around the Page/Folder insert? Not massively great either. Maybe having another queue to which you add the final upsert is the way to go?

Reply:
@spolu just realizing that we are updating pages parentURL which means they need to re-upserted or need to wait for the end before upserting any of them.

All of that to have the page in its folder. Is worth it?

Reply:
Parents can be updated separately if needed :+1: